Content.Server/GameObjects/Components/Weapon/Ranged/Barrels/BoltActionBarrelComponent.cs

```diff
@@ -0,0 +1,324 @@
+        }
+
+        protected override bool WeaponCanFire()
+        {
+            if (!base.WeaponCanFire())
+            {
+                return false;
+            }
+
+            return !BoltOpen && _chamberContainer.ContainedEntity != null;
+        }
+
+        private void Cycle(bool manual = false)
+        {
+            var chamberedEntity = _chamberContainer.ContainedEntity;
+            if (chamberedEntity != null)
+            {
+                _chamberContainer.Remove(chamberedEntity);
+                var ammoComponent = chamberedEntity.GetComponent<AmmoComponent>();
+                if (!ammoComponent.Caseless)
+                {
+                    EjectCasing(chamberedEntity);   
+                }
+            }
+
+            if (_spawnedAmmo.TryPop(out var next))
+            {
+                _ammoContainer.Remove(next);
+                _chamberContainer.Insert(next);
+            }
+
+            if (_unspawnedCount > 0)
+            {
+                _unspawnedCount--;
+                var ammoEntity = Owner.EntityManager.SpawnEntity(_fillPrototype, Owner.Transform.GridPosition);
+                _chamberContainer.Insert(ammoEntity);
+            }
+
+            if (_chamberContainer.ContainedEntity == null && manual)
+            {
+                BoltOpen = true;
+                if (ContainerHelpers.TryGetContainer(Owner, out var container))
+                {
+                    Owner.PopupMessage(container.Owner, Loc.GetString("Bolt opened"));
+                }
+            }
+
+            if (manual)
+            {
+                if (_soundCycle != null)
+                {
+                    var soundSystem = IoCManager.Resolve<IEntitySystemManager>().GetEntitySystem<AudioSystem>();
+                    soundSystem.PlayAtCoords(_soundCycle, Owner.Transform.GridPosition, AudioParams.Default.WithVolume(-2));
+                }
+            }
+            
+            Dirty();
+            UpdateAppearance();
+        }
+
+        public bool TryInsertBullet(IEntity user, IEntity ammo)
+        {
+            if (!ammo.TryGetComponent(out AmmoComponent ammoComponent))
+            {
+                return false;
+            }
+
+            if (ammoComponent.Caliber != _caliber)
+            {
+                Owner.PopupMessage(user, Loc.GetString("Wrong caliber"));
+                return false;
+            }
+
+            if (!BoltOpen)
+            {
+                Owner.PopupMessage(user, Loc.GetString("Bolt isn't open"));
+                return false;
+            }
+
+            if (_chamberContainer.ContainedEntity == null)
+            {
+                _chamberContainer.Insert(ammo);
+                if (_soundInsert != null)
+                {
+                    var soundSystem = IoCManager.Resolve<IEntitySystemManager>().GetEntitySystem<AudioSystem>();
+                    soundSystem.PlayAtCoords(_soundInsert, Owner.Transform.GridPosition, AudioParams.Default.WithVolume(-2));
```

Comment:
It looks like some of these calls should be using `PlayFromEntity` instead, if they're associated with `Owner`.